components/suggest/src/store.rs

```diff
@@ -0,0 +1,421 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    ops::Deref,
+    path::{Path, PathBuf},
+};
+
+use once_cell::sync::OnceCell;
+use remote_settings::{self, Attachment, GetItemsOptions, RemoteSettingsConfig, SortOrder};
+use serde_derive::*;
+
+use crate::{
+    db::{ConnectionType, SuggestDb, LAST_FETCH_META_KEY},
+    RemoteRecordId, RemoteSuggestion, Result, SuggestApiResult, Suggestion, SuggestionQuery,
+};
+
+const RS_COLLECTION: &str = "quicksuggest";
+
+/// The store is the entry point to the Suggest component. It incrementally
+/// fetches suggestions from the Remote Settings service, stores them in a local
+/// database, and returns them in response to user queries.
+///
+/// Your application should create a single store, and manage it as a singleton.
+/// The store is thread-safe, and supports concurrent fetches and ingests. We
+/// expect that your application will call `query` to show suggestions as the
+/// user types into the address bar, and periodically call `ingest` in the
+/// background to update the database with new suggestions from Remote Settings.
+///
+/// The store keeps track of the state needed to support incremental ingestion,
+/// but doesn't schedule the ingestion work itself, or decide how many
+/// suggestions to ingest at once. This is for two reasons:
+///
+/// 1. The primitives for scheduling background work vary between platforms. You
+///    might use an idle timer on Desktop, `WorkManager` on Android, or
+///    `BGTaskScheduler` on iOS.
+/// 2. Ingestion limits can change, depending on the platform and the needs of
+///    the application. A mobile device on a metered connection might want to
+///    request a small subset of the Suggest data and fetch the rest later,
+///    while a desktop on a fast link might request the entire dataset on first
+///    launch.
```

Comment:
💭 Some other ideas for using constraints:

* Imagine a future where we have "offline" location-based suggestions. We could have a scheme that chunks these suggestion by region in Remote Settings, and constrain the ingestion to just the chunk based on the user's current (coarse) location—so that we'll only download suggestions near where the user is, instead of downloading the _entire_ dataset and using up space to store suggestions for places they might never visit.
* If the user just installed Firefox, we might want to ingest an initial set of "higher-value" suggestions that we can show right away, and ingest the rest later. This would also require changing our Remote Settings schema, as I don't think records currently know what suggestions are in their attachments—but it's a thought for making Suggest more useful to first-time users.
* If the user is on cellular, we could similarly limit the amount of work we do. Android's `WorkManager` lets you schedule a task based on the network connection type, but we might still want to fetch a smaller set of suggestions even on cellular—just not a full fetch.